### ISSUE TITLE
Spark based benchmarks fail on OpenJ9

### ISSUE THREAD

**Issue by ceresek** [renaissance-benchmarks/renaissance#131]
Not sure if this is purely OpenJ9 issue or if we are involved with e.g. packaging, but the Spark based benchmarks fail on OpenJ9, here is an example dump from chi-square:

```
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
19/05/07 15:43:22 INFO SparkContext: Running Spark version 2.0.0
19/05/07 15:43:23 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
19/05/07 15:43:23 WARN SparkConf: In Spark 1.0 and later spark.local.dir will be overridden by the value set by the cluster manager (via SPARK_LOCAL_DIRS in mesos/standalone and LOCAL_DIRS in YARN).
19/05/07 15:43:23 INFO SecurityManager: Changing view acls to: root
19/05/07 15:43:23 INFO SecurityManager: Changing modify acls to: root
19/05/07 15:43:23 INFO SecurityManager: Changing view acls groups to: 
19/05/07 15:43:23 INFO SecurityManager: Changing modify acls groups to: 
19/05/07 15:43:23 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users  with view permissions: Set(root); groups with view permissions: Set(); users  with modify permissions: Set(root); groups with modify permissions: Set()
19/05/07 15:43:24 INFO Utils: Successfully started service 'sparkDriver' on port 39263.
19/05/07 15:43:24 INFO SparkEnv: Registering MapOutputTracker
19/05/07 15:43:24 INFO SparkEnv: Registering BlockManagerMaster
19/05/07 15:43:24 INFO DiskBlockManager: Created local directory at /var/tmp/tmp.4h1JqcRmg7/chi_square927351186687148580/blockmgr-92b98c5d-cb1e-4a9b-9859-877a4b344304
19/05/07 15:43:24 INFO MemoryStore: MemoryStore started with capacity 7.0 GB
19/05/07 15:43:24 INFO SparkEnv: Registering OutputCommitCoordinator
19/05/07 15:43:24 INFO Utils: Successfully started service 'SparkUI' on port 4040.
19/05/07 15:43:24 INFO SparkUI: Bound SparkUI to 0.0.0.0, and started at http://10.10.49.19:4040
19/05/07 15:43:24 INFO Executor: Starting executor ID driver on host localhost
19/05/07 15:43:24 INFO Utils: Successfully started service 'org.apache.spark.network.netty.NettyBlockTransferService' on port 34441.
19/05/07 15:43:24 INFO NettyBlockTransferService: Server created on 10.10.49.19:34441
19/05/07 15:43:24 INFO BlockManagerMaster: Registering BlockManager BlockManagerId(driver, 10.10.49.19, 34441)
19/05/07 15:43:24 INFO BlockManagerMasterEndpoint: Registering block manager 10.10.49.19:34441 with 7.0 GB RAM, BlockManagerId(driver, 10.10.49.19, 34441)
19/05/07 15:43:24 INFO BlockManagerMaster: Registered BlockManager BlockManagerId(driver, 10.10.49.19, 34441)
Error during tear-down: null
java.lang.NullPointerException
	at scala.collection.mutable.ArrayOps$ofRef$.length$extension(ArrayOps.scala:192)
	at scala.collection.mutable.ArrayOps$ofRef.length(ArrayOps.scala:192)
	at scala.collection.SeqLike$class.size(SeqLike.scala:106)
	at scala.collection.mutable.ArrayOps$ofRef.size(ArrayOps.scala:186)
	at scala.collection.mutable.Builder$class.sizeHint(Builder.scala:69)
	at scala.collection.mutable.ArrayBuilder.sizeHint(ArrayBuilder.scala:22)
	at scala.collection.TraversableLike$class.builder$1(TraversableLike.scala:230)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:233)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
	at org.renaissance.apache.spark.ChiSquare.tearDownAfterAll(ChiSquare.scala:98)
	at org.renaissance.RenaissanceBenchmark.runBenchmark(RenaissanceBenchmark.java:97)
	at org.renaissance.RenaissanceSuite$.$anonfun$main$2(renaissance-suite.scala:308)
	at org.renaissance.RenaissanceSuite$.$anonfun$main$2$adapted(renaissance-suite.scala:306)
	at org.renaissance.RenaissanceSuite$$$Lambda$93.000000005C8BA540.apply(Unknown Source)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.renaissance.RenaissanceSuite$.main(renaissance-suite.scala:306)
	at org.renaissance.RenaissanceSuite.main(renaissance-suite.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.renaissance.Launcher.main(Launcher.java:18)
Exception occurred in org.renaissance.apache.spark.ChiSquare@ec4bdca1: null
java.lang.ExceptionInInitializerError
	at java.lang.J9VMInternals.ensureError(J9VMInternals.java:148)
	at java.lang.J9VMInternals.recordInitializationFailure(J9VMInternals.java:137)
	at org.apache.commons.lang3.SerializationUtils.clone(SerializationUtils.java:88)
	at org.apache.spark.SparkContext$$anon$2.childValue(SparkContext.scala:336)
	at org.apache.spark.SparkContext$$anon$2.childValue(SparkContext.scala:332)
	at java.lang.ThreadLocal$ThreadLocalMap.<init>(ThreadLocal.java:391)
	at java.lang.ThreadLocal$ThreadLocalMap.<init>(ThreadLocal.java:298)
	at java.lang.ThreadLocal.createInheritedMap(ThreadLocal.java:255)
	at java.lang.Thread.initialize(Thread.java:356)
	at java.lang.Thread.<init>(Thread.java:324)
	at java.lang.Thread.<init>(Thread.java:117)
	at com.ibm.lang.management.internal.MemoryNotificationThread.<init>(MemoryNotificationThread.java:54)
	at com.ibm.lang.management.internal.ExtendedMemoryMXBeanImpl.<init>(ExtendedMemoryMXBeanImpl.java:50)
	at com.ibm.lang.management.internal.ExtendedMemoryMXBeanImpl.<clinit>(ExtendedMemoryMXBeanImpl.java:38)
	at com.ibm.java.lang.management.internal.ManagementUtils$Component.registerAll(ManagementUtils.java:728)
	at com.ibm.java.lang.management.internal.ManagementUtils$Metadata.<clinit>(ManagementUtils.java:1003)
	at com.ibm.java.lang.management.internal.ManagementUtils.getAllAvailableMXBeans(ManagementUtils.java:663)
	at java.lang.management.ManagementFactory$ServerHolder$1Registration.run(ManagementFactory.java:479)
	at java.lang.management.ManagementFactory$ServerHolder$1Registration.run(ManagementFactory.java:468)
	at java.security.AccessController.doPrivileged(AccessController.java:647)
	at java.lang.management.ManagementFactory$ServerHolder.<clinit>(ManagementFactory.java:505)
	at java.lang.management.ManagementFactory.getPlatformMBeanServer(ManagementFactory.java:247)
	at org.apache.spark.util.SizeEstimator$.getIsCompressedOops(SizeEstimator.scala:140)
	at org.apache.spark.util.SizeEstimator$.initialize(SizeEstimator.scala:112)
	at org.apache.spark.util.SizeEstimator$.<init>(SizeEstimator.scala:105)
	at org.apache.spark.util.SizeEstimator$.<clinit>(SizeEstimator.scala)
	at org.apache.spark.util.collection.SizeTracker$class.takeSample(SizeTracker.scala:78)
	at org.apache.spark.util.collection.SizeTracker$class.resetSamples(SizeTracker.scala:61)
	at org.apache.spark.util.collection.SizeTrackingVector.resetSamples(SizeTrackingVector.scala:25)
	at org.apache.spark.util.collection.SizeTracker$class.$init$(SizeTracker.scala:51)
	at org.apache.spark.util.collection.SizeTrackingVector.<init>(SizeTrackingVector.scala:26)
	at org.apache.spark.storage.memory.MemoryStore.putIteratorAsValues(MemoryStore.scala:199)
	at org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:919)
	at org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:910)
	at org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:866)
	at org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:910)
	at org.apache.spark.storage.BlockManager.putIterator(BlockManager.scala:700)
	at org.apache.spark.storage.BlockManager.putSingle(BlockManager.scala:1213)
	at org.apache.spark.broadcast.TorrentBroadcast.writeBlocks(TorrentBroadcast.scala:103)
	at org.apache.spark.broadcast.TorrentBroadcast.<init>(TorrentBroadcast.scala:86)
	at org.apache.spark.broadcast.TorrentBroadcastFactory.newBroadcast(TorrentBroadcastFactory.scala:34)
	at org.apache.spark.broadcast.BroadcastManager.newBroadcast(BroadcastManager.scala:56)
	at org.apache.spark.SparkContext.broadcast(SparkContext.scala:1370)
	at org.apache.spark.SparkContext$$anonfun$hadoopFile$1.apply(SparkContext.scala:984)
	at org.apache.spark.SparkContext$$anonfun$hadoopFile$1.apply(SparkContext.scala:981)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.SparkContext.withScope(SparkContext.scala:682)
	at org.apache.spark.SparkContext.hadoopFile(SparkContext.scala:981)
	at org.apache.spark.SparkContext$$anonfun$textFile$1.apply(SparkContext.scala:802)
	at org.apache.spark.SparkContext$$anonfun$textFile$1.apply(SparkContext.scala:800)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.SparkContext.withScope(SparkContext.scala:682)
	at org.apache.spark.SparkContext.textFile(SparkContext.scala:800)
	at org.renaissance.apache.spark.ChiSquare.loadData(ChiSquare.scala:66)
	at org.renaissance.apache.spark.ChiSquare.setUpBeforeAll(ChiSquare.scala:89)
	at org.renaissance.RenaissanceBenchmark.runBenchmark(RenaissanceBenchmark.java:74)
	at org.renaissance.RenaissanceSuite$.$anonfun$main$2(renaissance-suite.scala:308)
	at org.renaissance.RenaissanceSuite$.$anonfun$main$2$adapted(renaissance-suite.scala:306)
	at org.renaissance.RenaissanceSuite$$$Lambda$93.000000005C8BA540.apply(Unknown Source)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.renaissance.RenaissanceSuite$.main(renaissance-suite.scala:306)
	at org.renaissance.RenaissanceSuite.main(renaissance-suite.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.renaissance.Launcher.main(Launcher.java:18)
Caused by: java.lang.NullPointerException
	at org.apache.commons.lang3.SerializationUtils$ClassLoaderAwareObjectInputStream.<init>(SerializationUtils.java:300)
	at org.apache.commons.lang3.SerializationUtils.clone(SerializationUtils.java:88)
	at org.apache.spark.SparkContext$$anon$2.childValue(SparkContext.scala:336)
	at org.apache.spark.SparkContext$$anon$2.childValue(SparkContext.scala:332)
	at java.lang.ThreadLocal$ThreadLocalMap.<init>(ThreadLocal.java:391)
	at java.lang.ThreadLocal$ThreadLocalMap.<init>(ThreadLocal.java:298)
	at java.lang.ThreadLocal.createInheritedMap(ThreadLocal.java:255)
	at java.lang.Thread.initialize(Thread.java:356)
	at java.lang.Thread.<init>(Thread.java:324)
	at java.lang.Thread.<init>(Thread.java:231)
	at java.io.ClassCache$Reaper.<init>(ClassCache.java:221)
	at java.io.ClassCache$CreateReaperAction.run(ClassCache.java:212)
	at java.io.ClassCache$CreateReaperAction.run(ClassCache.java:201)
	at java.security.AccessController.doPrivileged(AccessController.java:647)
	at java.io.ClassCache.<init>(ClassCache.java:67)
	at java.io.ObjectInputStream.<clinit>(ObjectInputStream.java:346)
	... 69 more
```

**Comment by villazon** [comment]
On which version of the JDK this occurs?

**Comment by ceresek** [comment]
This was with:
```
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (build 1.8.0_212-b03)
Eclipse OpenJ9 VM (build openj9-0.14.0, JRE 1.8.0 Linux amd64-64-Bit Compressed References 20190417_286 (JIT enabled, AOT enabled)
OpenJ9   - bad1d4d06
OMR      - 4a4278e6
JCL      - 5590c4f818 based on jdk8u212-b03)
```

**Comment by DanHeidinga** [comment]
The OpenJ9 issue (https://github.com/eclipse/openj9/issues/5726) tracking the NPEs has been resolved and the fixes are available in the nightly builds.  They should be included in the `v0.14.2` refresh for Java 8 & 11 at AdoptOpenJDK in the next week or two

**Comment by ceresek** [comment]
@DanHeidinga Thank you! I've checked the benchmarks with OpenJDK8U-jdk_x64_linux_openj9_8u212b04_openj9-0.14.2.tar.gz and the crashes are gone (note to self: OpenJ9 requires setting -Xmx above the default of 512 MB for some benchmarks).
